src/Raven.Server/Documents/Replication/Outgoing/OutgoingMigrationReplicationHandler.cs

```diff
@@ -1,47 +1,70 @@
-﻿using System;
-using System.IO;
+﻿using System.IO;
+using Raven.Client.Documents.Replication.Messages;
 using Raven.Client.ServerWide.Commands;
 using Raven.Server.Documents.Replication.Senders;
+using Raven.Server.Documents.Sharding;
+using Raven.Server.Documents.Sharding.Operations;
+using Sparrow.Json.Parsing;
 using Sparrow.Logging;
+using Sparrow.Utils;
 
 namespace Raven.Server.Documents.Replication.Outgoing
 {
     public class OutgoingMigrationReplicationHandler : DatabaseOutgoingReplicationHandler
     {
+        private readonly ShardedDocumentDatabase _shardedDatabase;
         public readonly BucketMigrationReplication BucketMigrationNode;
-        public OutgoingMigrationReplicationHandler(ReplicationLoader parent, DocumentDatabase database, BucketMigrationReplication node, TcpConnectionInfo connectionInfo) : base(parent, database, node, connectionInfo)
+        public string LastChangeVectorInBucket = null;
+
+        public OutgoingMigrationReplicationHandler(ReplicationLoader parent, ShardedDocumentDatabase database, BucketMigrationReplication node, TcpConnectionInfo connectionInfo) : base(parent, database, node, connectionInfo)
         {
+            _shardedDatabase = database;
             BucketMigrationNode = node;
+            SuccessfulReplication += TryNotifySourceMigrationCompleted;
+        }
+
+        private void TryNotifySourceMigrationCompleted(DatabaseOutgoingReplicationHandler handler)
+        {
+            var current = handler as OutgoingMigrationReplicationHandler;
```

Comment:
if you are using safe cast and do not expect null then add a null guard

Reply:
done